### DISCUSSION TITLE
How to log dedicated server on windows.

### DISCUSSION THREAD

**Discussion by Shituation** [OpenTTD/OpenTTD#10648] [Q&A]
What title says.
I've tried:

openttd.exe -D >logfile.txt 2>&1
openttd.exe -D -d 4 >logfile.log

None seems to work. If I do that, I end up with an empty logfile. I need a verbose log (mostly user connections).

**Comment by erenes** [comment]
By default it won't work on windows because OpenTTD is compiled as a windows GUI application. To get the console output in the case of a dedicated server, it creates it's own console window and redirects stdin/stdout/stderr to that custom console window. 

This happens here: 
https://github.com/OpenTTD/OpenTTD/blob/master/src/os/windows/win32.cpp#L261

There is some weird stuff in Windows in this regard, to get really good first-party support for console apps we need to do two things: 
1) Use AttachConsole(ATTACH_PARENT_PROCESS) to attach to the console that spawns the process instead of AllocConsole
2) Compile OpenTTD as a console application (this means that when opening the application it will always open a console, so that's not ideal).

With those changes I got it to work: 
```
D:\Source\OpenTTD\build>.\debug\openttd.exe -D >logfile.txt 2>error.txt
unpause
pause
exit
```
![image](https://user-images.githubusercontent.com/3177915/232592867-27da1f0f-c7a8-477a-b8a4-5bf925a1362e.png)

Where error.txt contains: 
```
dbg: [net] Starting dedicated server, version 20230417-master-mf8b5661d28
dbg: [net] Starting network
dbg: [net] Initializing UDP listeners
dbg: [net] Network online, multiplayer available
dbg: [net] Detected broadcast addresses:
<redacted>
dbg: [net] Network revision name: 20230417-master-mf8b5661d28
dbg: [net] Initializing UDP listeners
dbg: [net] Initializing UDP listeners
dbg: [net] Listening on [::]:3979 (IPv6)
dbg: [net] Listening on 0.0.0.0:3979 (IPv4)
dbg: [net] Listening on [::]:3979 (IPv6)
dbg: [net] Listening on 0.0.0.0:3979 (IPv4)
dbg: [net] Generating map, please wait...
dbg: [net] Map generation percentage complete: 5
dbg: [net] Map generation percentage complete: 10
dbg: [net] Map generation percentage complete: 15
dbg: [net] Map generation percentage complete: 20
dbg: [net] Map generation percentage complete: 25
dbg: [net] Map generation percentage complete: 30
dbg: [net] Map generation percentage complete: 35
dbg: [net] Map generation percentage complete: 40
dbg: [net] Map generation percentage complete: 45
dbg: [net] Map generation percentage complete: 50
dbg: [net] Map generation percentage complete: 55
dbg: [net] Map generation percentage complete: 60
dbg: [net] Map generation percentage complete: 65
dbg: [net] Map generation percentage complete: 70
dbg: [net] Map generation percentage complete: 75
dbg: [net] Map generation percentage complete: 80
dbg: [net] Map generation percentage complete: 85
dbg: [net] Map generation percentage complete: 90
dbg: [net] Map generation percentage complete: 99
dbg: [net] Map generated, starting game
dbg: [net] Initializing UDP listeners
dbg: [net] Shutting down network
```

Because of the drawbacks of 2, this is not really feasible I think. Only doing 1 would also fix the output situation, but it causes issues with `stdin` ([see also](https://stackoverflow.com/questions/15952892/using-the-console-in-a-gui-app-in-windows-only-if-its-run-from-a-console))

I think the most logical solution is to add some logfile parameters to openttd to redirect stderr and stdout instead of distributing a console app.

**Reply by erenes** [reply]
I have put my WIP code as an example here: 
https://github.com/erenes/OpenTTD/commit/e89d29ebdc1db76fafd34c865b33bdff0386db14

I don't recommend running it.

**Comment by rubidium42** [comment]
Technically there is a simpler solution, as it's just some bit in the binary that determines the behaviour. There is even a tool to convert an exe between 'GUI-mode' and 'console-mode': https://www.tt-forums.net/viewtopic.php?t=39826
Only problem is that the link does not work anymore. @glx22, do you still have that executable and can/do we share that somewhere?

**Comment by glx22** [comment]
Yes I still have the files
[convert.zip](https://github.com/OpenTTD/OpenTTD/files/11256385/convert.zip)

**Reply by erenes** [reply]
Yup, much easier and no code changes required :)

Confirmed it works by: 
1) Placing convert.exe next to openttd.exe
2) Running convert.exe from the CLI
3) Running `openttd.exe -D >logfile.txt 2>error.txt`, which logs as expected.

**Reply by glx22** [reply]
And you can run convert.exe again to switch back to GUI.

**Reply by Shituation** [reply]
`failed to open openttd.exe for read`

- Game exe is on the same folder than the patch. Game is installed on C:\openttd\ 
- All permissions granted. 
- Tried both game versions win32 and win64. 
- Game version 13.1.
